File: src/services/oauth-manager.ts
import { logServerEvent } from '../logger';

export interface OAuthConfig {
  tokenEndpoint: string;
  clientId: string;
  clientSecret: string;
  scope?: string;
  grantType?: string; // default: 'client_credentials'
  refreshBufferMs?: number; // default: 5 minutes before expiry
}

export interface OAuthToken {
  accessToken: string;
  expiresAt: number; // Unix timestamp in ms
  tokenType: string;
}

/**
 * OAuth Token Manager
 * Handles automatic token fetching and refreshing for target endpoint authentication
 */
export class OAuthManager {
  private config: OAuthConfig;
  private currentToken: OAuthToken | null = null;
  private refreshTimer: NodeJS.Timeout | null = null;
  private isRefreshing: boolean = false;
  private refreshPromise: Promise<void> | null = null;

  constructor(config: OAuthConfig) {
    this.config = {
      grantType: 'client_credentials',
      refreshBufferMs: 5 * 60 * 1000, // 5 minutes default
      ...config,
    };
  }

  /**
   * Initialize the OAuth manager and fetch the first token
   */
  async initialize(): Promise<void> {
    logServerEvent('info', 'Initializing OAuth token manager', {
      tokenEndpoint: this.config.tokenEndpoint,
      clientId: this.config.clientId,
    });

    await this.refreshToken();
  }

  /**
   * Get a valid access token, refreshing if necessary
   */
  async getAccessToken(): Promise<string> {
    // If no token, fetch one
    if (!this.currentToken) {
      await this.refreshToken();
      return this.currentToken!.accessToken;
    }

    // If token is about to expire or expired, refresh
    const now = Date.now();
    const bufferMs = this.config.refreshBufferMs || 5 * 60 * 1000;

    if (now >= this.currentToken.expiresAt - bufferMs) {
      // If already refreshing, wait for that to complete
      if (this.isRefreshing && this.refreshPromise) {
        await this.refreshPromise;
      } else {
        await this.refreshToken();
      }
    }

    return this.currentToken!.accessToken;
  }

  /**
   * Fetch a new access token from the OAuth endpoint
   */
  private async refreshToken(): Promise<void> {
    // Prevent concurrent refresh attempts
    if (this.isRefreshing) {
      if (this.refreshPromise) {
        return this.refreshPromise;
      }
    }

    this.isRefreshing = true;
    this.refreshPromise = this.doRefresh();

    try {
      await this.refreshPromise;
    } finally {
      this.isRefreshing = false;
      this.refreshPromise = null;
    }
  }

  /**
   * Actual token refresh logic
   */
  private async doRefresh(): Promise<void> {
    try {
      logServerEvent('info', 'Fetching new OAuth token', {
        endpoint: this.config.tokenEndpoint,
      });

      // Prepare OAuth request body
      const body = new URLSearchParams({
        grant_type: this.config.grantType || 'client_credentials',
        client_id: this.config.clientId,
        client_secret: this.config.clientSecret,
      });

      // Add scope if provided
      if (this.config.scope) {
        body.append('scope', this.config.scope);
      }

      // Make OAuth token request
      const response = await fetch(this.config.tokenEndpoint, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/x-www-form-urlencoded',
        },
        body: body.toString(),
      });

      if (!response.ok) {
        const errorText = await response.text();
        throw new Error(`OAuth token request failed: ${response.status} ${errorText}`);
      }

      const data = await response.json();

      // Validate response
      if (!data.access_token) {
        throw new Error('OAuth response missing access_token field');
      }

      // Calculate expiry time
      // expires_in is typically in seconds
      const expiresInMs = (data.expires_in || 3600) * 1000;
      const expiresAt = Date.now() + expiresInMs;

      // Store the token
      this.currentToken = {
        accessToken: data.access_token,
        expiresAt,
        tokenType: data.token_type || 'Bearer',
      };

      logServerEvent('info', 'OAuth token acquired successfully', {
        expiresIn: data.expires_in,
        expiresAt: new Date(expiresAt).toISOString(),
        tokenType: this.currentToken.tokenType,
      });

      // Schedule next refresh
      this.scheduleRefresh();
    } catch (error) {
      logServerEvent('error', 'Failed to refresh OAuth token', {
        error: error instanceof Error ? error.message : String(error),
      });
      throw error;
    }
  }

  /**
   * Schedule the next token refresh
   */
  private scheduleRefresh(): void {
    // Clear existing timer
    if (this.refreshTimer) {
      clearTimeout(this.refreshTimer);
    }

    if (!this.currentToken) {
      return;
    }

    // Calculate when to refresh (with buffer)
    const now = Date.now();
    const bufferMs = this.config.refreshBufferMs || 5 * 60 * 1000;
    const refreshAt = this.currentToken.expiresAt - bufferMs;
    const delay = Math.max(0, refreshAt - now);

    logServerEvent('info', 'Scheduled next OAuth token refresh', {
      currentTime: new Date(now).toISOString(),
      refreshAt: new Date(refreshAt).toISOString(),
      delayMs: delay,
      delayMinutes: Math.round(delay / 60000),
    });

    this.refreshTimer = setTimeout(() => {
      this.refreshToken().catch((error) => {
        logServerEvent('error', 'Scheduled OAuth token refresh failed', {
          error: error instanceof Error ? error.message : String(error),
        });
        // Try again in 1 minute on failure
        setTimeout(() => {
          this.refreshToken();
        }, 60000);
      });
    }, delay);
  }

  /**
   * Get the current token info (for debugging/monitoring)
   */
  getTokenInfo(): { hasToken: boolean; expiresAt?: string; isExpired?: boolean } {
    if (!this.currentToken) {
      return { hasToken: false };
    }

    const now = Date.now();
    return {
      hasToken: true,
      expiresAt: new Date(this.currentToken.expiresAt).toISOString(),
      isExpired: now >= this.currentToken.expiresAt,
    };
  }

  /**
   * Force a token refresh (useful for testing or error recovery)
   */
  async forceRefresh(): Promise<void> {
    logServerEvent('info', 'Forcing OAuth token refresh');
    this.currentToken = null;
    await this.refreshToken();
  }

  /**
   * Cleanup on shutdown
   */
  destroy(): void {
    if (this.refreshTimer) {
      clearTimeout(this.refreshTimer);
      this.refreshTimer = null;
    }
    this.currentToken = null;
    logServerEvent('info', 'OAuth manager destroyed');
  }
}

// Singleton instance
let oauthManagerInstance: OAuthManager | null = null;

/**
 * Initialize the global OAuth manager
 */
export async function initializeOAuth(config: OAuthConfig): Promise<void> {
  if (oauthManagerInstance) {
    oauthManagerInstance.destroy();
  }

  oauthManagerInstance = new OAuthManager(config);
  await oauthManagerInstance.initialize();
}

/**
 * Get the current OAuth manager instance
 */
export function getOAuthManager(): OAuthManager | null {
  return oauthManagerInstance;
}

/**
 * Check if OAuth is configured and initialized
 */
export function isOAuthEnabled(): boolean {
  return oauthManagerInstance !== null;
}

File: oauth_manager.py
"""OAuth 2.0 token manager with automatic refresh."""

import time
import logging
import requests
import threading
from typing import Optional, Dict, Any

logger = logging.getLogger(__name__)


class OAuthManager:
    """Manages OAuth token fetching and auto-refresh."""

    def __init__(
        self,
        token_endpoint: str,
        client_id: str,
        client_secret: str,
        scope: Optional[str] = None,
        refresh_buffer_minutes: int = 5
    ):
        self.token_endpoint = token_endpoint
        self.client_id = client_id
        self.client_secret = client_secret
        self.scope = scope
        self.refresh_buffer_seconds = refresh_buffer_minutes * 60

        self._access_token: Optional[str] = None
        self._expires_at: Optional[float] = None
        self._refresh_timer: Optional[threading.Timer] = None
        self._lock = threading.Lock()

    def get_token(self) -> Optional[str]:
        """Get a valid access token (refreshes if needed)."""
        with self._lock:
            # Check if we need to refresh
            if not self._access_token or self._needs_refresh():
                self._fetch_token()

            return self._access_token

    def _needs_refresh(self) -> bool:
        """Check if token needs to be refreshed."""
        if not self._expires_at:
            return True

        # Refresh if we're within buffer time of expiry
        time_until_expiry = self._expires_at - time.time()
        return time_until_expiry <= self.refresh_buffer_seconds

    def _fetch_token(self) -> None:
        """Fetch a new OAuth token."""
        try:
            logger.info(f"Fetching OAuth token from {self.token_endpoint}")

            data = {
                'grant_type': 'client_credentials',
            }

            if self.scope:
                data['scope'] = self.scope

            # Try with Basic Auth first (some OAuth servers prefer this)
            from requests.auth import HTTPBasicAuth
            auth = HTTPBasicAuth(self.client_id, self.client_secret)

            logger.debug(f"OAuth request: grant_type=client_credentials, scope={self.scope}")

            response = requests.post(
                self.token_endpoint,
                data=data,
                auth=auth,
                headers={'Content-Type': 'application/x-www-form-urlencoded'},
                timeout=30
            )

            # If Basic Auth fails with 400, try with credentials in body
            if response.status_code == 400:
                logger.warning("Basic Auth failed, trying with credentials in request body...")
                data['client_id'] = self.client_id
                data['client_secret'] = self.client_secret

                response = requests.post(
                    self.token_endpoint,
                    data=data,
                    headers={'Content-Type': 'application/x-www-form-urlencoded'},
                    timeout=30
                )

            if not response.ok:
                error_detail = ""
                try:
                    error_data = response.json()
                    error_detail = f": {error_data}"
                except:
                    error_detail = f": {response.text}"

                logger.error(f"OAuth token request failed with {response.status_code}{error_detail}")
                raise Exception(f"OAuth request failed: {response.status_code}{error_detail}")

            response.raise_for_status()
            token_data = response.json()

            self._access_token = token_data.get('access_token')
            expires_in = token_data.get('expires_in', 3600)
            self._expires_at = time.time() + expires_in

            logger.info(f"✓ OAuth token obtained (expires in {expires_in}s)")

            # Schedule next refresh
            self._schedule_refresh()

        except Exception as e:
            logger.error(f"Failed to fetch OAuth token: {e}")
            self._access_token = None
            self._expires_at = None

    def _schedule_refresh(self) -> None:
        """Schedule the next token refresh."""
        if self._refresh_timer:
            self._refresh_timer.cancel()

        if not self._expires_at:
            return

        # Schedule refresh before expiry
        time_until_refresh = max(
            (self._expires_at - time.time()) - self.refresh_buffer_seconds,
            0
        )

        self._refresh_timer = threading.Timer(time_until_refresh, self._refresh_token)
        self._refresh_timer.daemon = True
        self._refresh_timer.start()

    def _refresh_token(self) -> None:
        """Background token refresh."""
        logger.info("Refreshing OAuth token...")
        with self._lock:
            self._fetch_token()

    def get_token_info(self) -> Dict[str, Any]:
        """Get token information for debugging."""
        return {
            'has_token': bool(self._access_token),
            'expires_at': self._expires_at,
            'time_until_expiry': self._expires_at - time.time() if self._expires_at else None,
        }

    def destroy(self) -> None:
        """Clean up resources."""
        if self._refresh_timer:
            self._refresh_timer.cancel()
